src/routers/account.js

```diff
@@ -494,4 +499,150 @@ router.get('/miieditor', async (request, response) => {
+		const price = prices.find(price => price.product === product.id);
+
+		return {
+			price_id: price.id,
+			thumbnail: product.images[0],
+			name: product.name,
+			description: product.description,
+			price: (price.unit_amount / 100).toLocaleString('en-US', { style: 'currency', currency: 'USD' }),
+		};
+	});
+
+	response.render('account/upgrade', renderData);
+});
+
+router.post('/checkout/:priceId', async (request, response) => {
+	// Verify the user is logged in
+	if (!request.cookies.access_token || !request.cookies.refresh_token || !request.cookies.ph) {
+		return response.redirect('/account/login');
+	}
+
+	// Attempt to get user data
+	let apiResponse = await util.apiGetRequest('/v1/user', {
+		'Authorization': `${request.cookies.token_type} ${request.cookies.access_token}`
+	});
+
+	if (apiResponse.statusCode !== 200) {
+		// Assume expired, refresh and retry request
+		apiResponse = await util.apiPostGetRequest('/v1/login', {}, {
+			refresh_token: request.cookies.refresh_token,
+			grant_type: 'refresh_token'
+		});
+
+		if (apiResponse.statusCode !== 200) {
+			return response.redirect('/account/login');
+		}
+
+		const tokens = apiResponse.body;
+
+		response.cookie('refresh_token', tokens.refresh_token, { domain: '.pretendo.network' });
+		response.cookie('access_token', tokens.access_token, { domain: '.pretendo.network' });
+		response.cookie('token_type', tokens.token_type, { domain: '.pretendo.network' });
+
+		apiResponse = await util.apiGetRequest('/v1/user', {
+			'Authorization': `${tokens.token_type} ${tokens.access_token}`
+		});
+	}
+
+	// If still failed, something went horribly wrong
+	if (apiResponse.statusCode !== 200) {
+		return response.redirect('/account/login');
+	}
+
+	// Set user account info to render data
+	const account = apiResponse.body;
+	const pid = account.pid;
+
+	let customer;
+	const { data: searchResults } = await await stripe.customers.search({
+		query: `metadata['pnid_pid']:'${pid}'`
+	});
+
+	if (searchResults.length !== 0) {
+		customer = searchResults[0];
+	} else {
+		customer = await stripe.customers.create({
+			metadata: {
+				pnid_pid: pid
+			}
+		});
+	}
+
+	const priceId = request.params.priceId;
+
+	const session = await stripe.checkout.sessions.create({
+		line_items: [
+			{
+				price: priceId,
+				quantity: 1,
+			},
+		],
+		customer: customer.id,
+		mode: 'subscription',
+		success_url: `${config.http.base_url}/account?upgrade_success=true`,
+		cancel_url: `${config.http.base_url}/account?upgrade_success=false`
+	});
+
+	response.redirect(303, session.url);
+});
+
+router.post('/stripe-wh', express.raw({ type: 'application/json' }), async (request, response) => {
+	const stripeSignature = request.headers['stripe-signature'];
+	let event;
+
+	try {
+		event = stripe.webhooks.constructEvent(request.body, stripeSignature, config.stripe.webhook_secret);
+	} catch (err) {
+		return response.status(400).send(`Webhook Error: ${err.message}`);
+	}
```

Comment:
On boot, you should use stripe's event list endpoint to recover any potentionally missed data. (Filter on unsuccessful delivery). This will make the service self healing.

https://stripe.com/docs/api/events